Mathlib/Combinatorics/SimpleGraph/Subgraph.lean

```diff
@@ -1150,6 +1150,28 @@ theorem deleteVerts_inter_verts_set_right_eq :
 
 end DeleteVerts
 
+/-! ### Of function -/
+
+/-- Given a function, construct the Subgraph where each element is connected
+  to its value under the function --/
+@[simps]
+def ofFunction {u : Set V} (f : V → V) (h : ∀ v ∈ u, G.Adj v (f v)) : Subgraph G where
```

Comment:
Looks like this be better defined as a graph rather than a subgraph? Then your current `Subgraph.ofFunction f h` becomes `SimpleGraph.ofFunction f ⊓ G`

Reply:
```
@[simps]
def ofFunction {u : Set V} (f : V → V) (h : ∀ v ∈ u, f v ≠ v) : SimpleGraph V where
  Adj v w := v ∈ u ∧ f v = w ∨ w ∈ u ∧ f w = v
```

```
def ofFunction {u : Set V} (f : V → V) (h : ∀ v ∈ u, G.Adj v (f v)) : Subgraph G :=
  G.toSubgraph (SimpleGraph.ofFunction f (fun v hv ↦ (h v hv).ne.symm ) ⊓ G) inf_le_right
```

I prepared this [pushed here](https://github.com/leanprover-community/mathlib4/tree/PO_subgraphOfFunction_sidequest) but in this configuration the resulting subgraph has the universe as matching, and hence is not `IsMatching`.

Any thoughts on how to properly tie this together?

Reply:
Sorry, I don't understand your message. Can you expand?

Reply:
My bad, I meant to say: the resulting subgraph has `Set.univ` as `verts`, so it's not `IsMatching`, so we'd need to in addition restrict to `u ∪ f '' u`. If somehow add that on, do you still think it's better than the direct definition?

In general I have kind of been pondering if it makes sense to see if either it's worthwhile to

- Introduce SimpleSubgraph of `G : SimpleGraph V` just to be a `H : SimpleGraph V` packaged with a proof that `G ≤ H`. - Maybe even replace Subgraph with the above definition. I understand there is a semantic difference in that you can't represent the inclusion or exclusion of isolated nodes, but I'm wondering how much this is actually used.

Any thoughts on that?

Reply:
> My bad, I meant to say: the resulting subgraph has `Set.univ` as `verts`, so it's not `IsMatching`, so we'd need to in addition restrict to `u ∪ f '' u`. If somehow add that on, do you still think it's better than the direct definition?

Can you show me where you are going to use `ofFunction`?

> Any thoughts on that?

I have been moving away from design decisions about graphs because it hurts my head 😁

Reply:
[Right here](https://github.com/pimotte/TutteLean/blob/main/TutteLean%2FPartNew.lean#L176-L178)

The situation here is the case where the graph decomposes into cliques after removing all universal vertices. I'm using this construction to pick some representatives from odd (clique) components and match them to universal vertices. (After that it's just even components that we can match internally)

Reply:
I see. I think you took a wrong turn here. In your code, rather than turning the assumption `{c : G.deleteUniversalVerts.coe.ConnectedComponent | Odd (c.supp.ncard)}.ncard ≤ G.universalVerts.ncard` into a partial function `f` then a total function `g`, you could directly show that `{c : G.deleteUniversalVerts.coe.ConnectedComponent | Odd (c.supp.ncard)}.ncard ≤ G.universalVerts.ncard` implies that there is a matching such that each element from the first set to the second, in the sense that each element of `{c : G.deleteUniversalVerts.coe.ConnectedComponent | Odd (c.supp.ncard)}` contains an element that's matched and each element of `G.universalVerts` is matched (secretly, you're replacing `G.universalVerts` by `{{x] | x ∈ G.universalVerts}`, which has the same size). This is a special case of Hall's marriage theorem!

As supporting evidence for my claim above that you took a wrong turn, note that you are not using `f` and `g` for anything besides defining the matching.

Reply:
Ah, yes, that makes sense. I'll explore that path and withdraw this PR. Thanks!